pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -0,0 +1,311 @@
+		return m.tenantIDs, nil, nil
+	}
+
+	// ensure the name of a retained tenant id label gets handled under the
+	// original label name
+	if name == originalDefaultTenantLabel {
+		name = defaultTenantLabel
+	}
+
+	return m.mergeDistinctStringSlice(func(q storage.Querier) ([]string, storage.Warnings, error) {
+		return q.LabelValues(name)
+	})
+}
+
+// LabelNames returns all the unique label names present in the underlying
+// queriers. It also adds the defaultTenantLabel and if present in the original
+// results the originalDefaultTenantLabel
+func (m *mergeQuerier) LabelNames() ([]string, storage.Warnings, error) {
+	labelNames, warnings, err := m.mergeDistinctStringSlice(func(q storage.Querier) ([]string, storage.Warnings, error) {
+		return q.LabelNames()
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// check if the tenant label exists in the original result
+	var tenantLabelExists bool
+	labelPos := sort.SearchStrings(labelNames, defaultTenantLabel)
+	if labelPos < len(labelNames) && labelNames[labelPos] == defaultTenantLabel {
+		tenantLabelExists = true
+	}
+
+	labelToAdd := defaultTenantLabel
+
+	// if defaultTenantLabel already exists, we need to add the
+	// originalDefaultTenantLabel
+	if tenantLabelExists {
+		labelToAdd = originalDefaultTenantLabel
+		labelPos = sort.SearchStrings(labelNames, labelToAdd)
+	}
+
+	// insert label at the correct position
+	labelNames = append(labelNames, "")
+	copy(labelNames[labelPos+1:], labelNames[labelPos:])
+	labelNames[labelPos] = labelToAdd
+
+	return labelNames, warnings, nil
+}
+
+type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
+
+// mergeDistinctStringSlice is aggregating results from stringSliceFunc calls
+// on a querier. It removes duplicates and sorts the result. It doesn't require
+// the output of the stringSliceFunc to be sorted, as results of LabelValues
+// are not sorted.
+//
+// TODO: Consider running stringSliceFunc calls concurrently
+func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+	var warnings storage.Warnings
+	resultMap := make(map[string]struct{})
+	for pos, tenantID := range m.tenantIDs {
+		result, resultWarnings, err := f(m.queriers[pos])
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, e := range result {
+			resultMap[e] = struct{}{}
+		}
+		for _, w := range resultWarnings {
+			warnings = append(warnings, fmt.Errorf("error querying tenant id %s: %w", tenantID, w))
+		}
+	}
+
+	var result = make([]string, 0, len(resultMap))
+	for e := range resultMap {
+		result = append(result, e)
+	}
+	sort.Strings(result)
+	return result, warnings, nil
+}
+
+// Close releases the resources of the Querier.
+func (m *mergeQuerier) Close() error {
+	errs := tsdb_errors.NewMulti()
+	for pos, tenantID := range m.tenantIDs {
+		if err := m.queriers[pos].Close(); err != nil {
+			errs.Add(errors.Wrapf(m.queriers[pos].Close(), "failed to close querier for tenant id %s", tenantID))
```

Comment:
This is not correct. You're closing twice.

Reply:
Right. `if err := ...` part should be removed. Both `Wrapf` and `Add` already handle case of `nil` error.

Reply:
Thanks very good catch and a lapse that can be dangerous: 42663d136